### PR TITLE
Changes that allow the pipeline to run within a shifter container.

### PR DESCRIPTION
**Pull request by tskisner**
desi_pipe now generates slurm scripts which use a specified docker image containing the desi software and dependencies.

### CODE REVIEW

**Comment by tskisner** [comment]
Obviously don't merge yet- will fix the small problem in the unit test.

File: py/desispec/parallel.py [hunk 1]
```diff
@@ -147,41 +157,58 @@ def dist_discrete(worksizes, workers, id, pow=1.0):
     return dist[id]
 
 
-
 @contextmanager
-def stdouterr_redirected(to=os.devnull, comm=None):
+def stdouterr_redirected(to=None, comm=None):
```
Comment (sbailey) [inline py/desispec/parallel.py]: Please update docstring to clarify whether `to` is a file name, open file handle, either, etc., what happens in the default case, and also what `comm` is (I know, but future developers may not...)

File: py/desispec/pipeline/plan.py [hunk 2]
```diff
@@ -60,7 +60,7 @@ def select_nights(allnights, nightstr):
     return nights
 
 
-def create_prod(nightstr=None):
+def create_prod(nightstr=None, extra={}):
```
Comment (sbailey) [inline py/desispec/pipeline/plan.py]: Please document what `extra` is and how it is used.

File: py/desispec/pipeline/run.py [hunk 3]
```diff
@@ -619,7 +619,7 @@ def nersc_job(path, logroot, desisetup, commands, nodes=1, \
     return
 
 
-def nersc_shifter_job(path, img, specdata, specredux, desimodel, logroot, desisetup, commands, nodes=1, \
+def nersc_shifter_job(path, img, specdata, specredux, desiroot, logroot, desisetup, commands, nodes=1, \
```
Comment (sbailey) [inline py/desispec/pipeline/run.py]: Needs docstring

File: py/desispec/pipeline/task.py [hunk 4]
```diff
@@ -931,15 +949,19 @@ def get_worker(step, name, opts):
     return worker
 
 
-def default_options():
+def default_options(extra={}):
```
Comment (sbailey) [inline py/desispec/pipeline/task.py]: Needs docstring

File: py/desispec/scripts/pipe_prod.py [hunk 5]
```diff
@@ -113,7 +113,8 @@ def compute_step_slurm(setupfile, first, last, specs, night,
     return nersc_path
 
 
-def compute_step_shifter(img, specdata, specredux, desimodel, setupfile, first, last, specs, night, 
+def compute_step_shifter(img, specdata, specredux, desiroot, setupfile,
```
Comment (sbailey) [inline py/desispec/scripts/pipe_prod.py]: Needs docstring

File: py/desispec/scripts/pipe_prod.py [hunk 6]
```diff
@@ -145,17 +146,17 @@ def compute_step_shifter(img, specdata, specredux, desimodel, setupfile, first,
...
 
 
-def compute_step(img, specdata, specredux, desimodel, setupfile, first, last, specs, night, 
-    ntask, taskproc, shell_mpi_run, shell_maxcores, shell_threads, nersc_maxnodes,
-    nersc_nodecores, nersc_threads, nersc_mp, nersc_queue_thresh,
-    queue="debug", minutes=30
+def compute_step(img, specdata, specredux, desiroot, setupfile, 
+    first, last, specs, night, ntask, taskproc, shell_mpi_run, shell_maxcores,
+    shell_threads, nersc_maxnodes, nersc_nodecores, nersc_threads, nersc_mp, 
+    nersc_queue_thresh, queue="debug", minutes=30
```
Comment (sbailey) [inline py/desispec/scripts/pipe_prod.py]: Needs docstring

**Pull request merged**